org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGenerator.java

```diff
@@ -10,178 +10,151 @@
         archetype.getArtifactId()));
     if(location == null) {
       // if the project should be created in the workspace, figure out the path
       location = workspaceRoot.getLocation();
     }
-
-    List<IProject> createdProjects = new ArrayList<>();
-
+    File basedir = location.toFile();
+    if(basedir == null || (!basedir.mkdirs() && !basedir.isDirectory())) {
+      throw new CoreException(Status.error(Messages.ProjectConfigurationManager_error_failed));
+    }
+    //See https://maven.apache.org/archetype/maven-archetype-plugin/generate-mojo.html
+    Properties userProperties = new Properties(properties);
+    userProperties.setProperty("archetypeGroupId", archetype.getGroupId());
+    userProperties.setProperty("archetypeArtifactId", archetype.getArtifactId());
+    userProperties.setProperty("archetypeVersion", archetype.getVersion());
+    String repository = archetype.getRepository();
+    if(repository != null) {
+      userProperties.setProperty("archetypeRepository", repository);
+    }
+    userProperties.setProperty("groupId", groupId);
+    userProperties.setProperty("artifactId", artifactId);
+    userProperties.setProperty("version", version);
+    userProperties.setProperty("package", javaPackage);
+    userProperties.setProperty("outputDirectory", basedir.getAbsolutePath());
+    String projectFolder = location.append(artifactId).toFile().getAbsolutePath();
+    File emptyPom = getEmptyPom(basedir);
     try {
-
-      Artifact artifact = resolveArchetype(archetype, monitor);
-
-      ArchetypeGenerationRequest request = new ArchetypeGenerationRequest() //
-          .setTransferListener(((MavenImpl) maven).createTransferListener(monitor)) //
-          .setArchetypeGroupId(artifact.getGroupId()) //
-          .setArchetypeArtifactId(artifact.getArtifactId()) //
-          .setArchetypeVersion(artifact.getVersion()) //
-          .setArchetypeRepository(archetype.getRepository()) //
-          .setGroupId(groupId) //
-          .setArtifactId(artifactId) //
-          .setVersion(version) //
-          .setPackage(javaPackage) // the model does not have a package field
-          .setLocalRepository(maven.getLocalRepository()) //
-          .setRemoteArtifactRepositories(maven.getArtifactRepositories(true)).setProperties(properties)
-          .setOutputDirectory(location.toPortableString());
-
-      ArchetypeGenerationResult result = Adapters.adapt(archetype, org.apache.maven.archetype.ArchetypeManager.class)
-          .generateProjectFromArchetype(request);
-
-      Exception cause = result.getCause();
-      if(cause != null) {
-        String msg = NLS.bind(Messages.ProjectConfigurationManager_error_unable_archetype, archetype);
-        log.error(msg, cause);
-        throw new CoreException(Status.error(msg, cause));
+      String goals = "-U org.apache.maven.plugins:maven-archetype-plugin:2.4:generate";
+      if(emptyPom != null) {
+        goals += " -f " + emptyPom.getAbsolutePath();
```

Comment:
What do you mean by 'new'? This is to prevent m2e to first generate the code at a temporary place and then move it around, without the empty pom you will get higher build times and can't create a module of a parent that has pom-errors and at least there is a test for that (don't know if it uses intentionally an incomplete module pom).

Reply:
By the way there are currently two restrictions one might report to maven and submit a patch, but this would then take some time until it is released:

1. If running `mvn` inside an empty folder / without a pom, it uses a "no-pom build", one can then force a different location of the pom by -f but if the folder contains no pom I can't force to ignore that file.
2. maven-archetype always generates the archetype at the folder `$outputDir/$artifactId` while here it would be good if we can just generate it in the current folder (what would be empty anyways).